### PR TITLE
Fix escaping special characters for data set entries.  

### PR DESCRIPTION
**Pull request by richp405**
##### SUMMARY

Bugfix to add escaping to dsnames for job_submit (and other ds-based functions)

** This is a partial fix... zos core will send properly escaped strings with this, but some zoau functions need to process the escapes, as well.  This is in 1.3.x (not yet released).

Addresses #642, #1011, #724 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
- zos_data_set
- zos_job_submit
- routines that call zos_data_set
- escaping routine added to modules/job.py
-

### CODE REVIEW

File: plugins/module_utils/job.py
Comment (fernandofloresg) [inline plugins/module_utils/job.py:1]: Please do a sanity execution on this I got a bunch of errors: 

```
(AC2.14PY3.9) fernandoflores@Fernandos-MacBook-Pro ibm_zos_core % ./ac --ac-sanity --python 3.11
-----------------------------------------------------------------------
Running ansible-test with managed python virtual environment: /Users/fernandoflores/Documents/IBM/Repos/ibm_zos_core/venv/venv-2.15.
-----------------------------------------------------------------------
Running sanity test "action-plugin-docs"
Running sanity test "ansible-doc"
Running sanity test "changelog"
Running sanity test "compile" on Python 3.10
Running sanity test "empty-init"
Running sanity test "future-import-boilerplate"
Running sanity test "ignores"
Running sanity test "import" on Python 3.10
ERROR: Found 5 import issue(s) on python 3.10 which need to be resolved:
ERROR: plugins/module_utils/job.py:426:17: traceback: SyntaxError: invalid escape sequence '\$'
ERROR: plugins/modules/zos_data_set.py:676:0: traceback: SyntaxError: invalid escape sequence '\$' (at plugins/module_utils/job.py:426:17)
ERROR: plugins/modules/zos_job_output.py:417:0: traceback: SyntaxError: invalid escape sequence '\$' (at plugins/module_utils/job.py:426:17)
ERROR: plugins/modules/zos_job_query.py:258:0: traceback: SyntaxError: invalid escape sequence '\$' (at plugins/module_utils/job.py:426:17)
ERROR: plugins/modules/zos_job_submit.py:610:0: traceback: SyntaxError: invalid escape sequence '\$' (at plugins/module_utils/job.py:426:17)
See documentation for help: https://docs.ansible.com/ansible-core/2.15/dev_guide/testing/sanity/import.html
Running sanity test "line-endings"
Running sanity test "metaclass-boilerplate"
Running sanity test "no-assert"
Running sanity test "no-basestring"
Running sanity test "no-dict-iteritems"
Running sanity test "no-dict-iterkeys"
Running sanity test "no-dict-itervalues"
Running sanity test "no-get-exception"
Running sanity test "no-illegal-filenames"
Running sanity test "no-main-display"
Running sanity test "no-smart-quotes"
Running sanity test "no-unicode-literals"
Running sanity test "pep8"
ERROR: Found 30 pep8 issue(s) which need to be resolved:
ERROR: plugins/module_utils/job.py:398:15: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:398:23: E202: whitespace before ')'
ERROR: plugins/module_utils/job.py:409:15: E275: missing whitespace after keyword
ERROR: plugins/module_utils/job.py:409:16: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:409:31: E202: whitespace before ')'
ERROR: plugins/module_utils/job.py:418:23: E275: missing whitespace after keyword
ERROR: plugins/module_utils/job.py:418:24: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:418:51: E202: whitespace before ')'
ERROR: plugins/module_utils/job.py:426:19: W605: invalid escape sequence '\$'
ERROR: plugins/module_utils/job.py:436:23: E275: missing whitespace after keyword
ERROR: plugins/module_utils/job.py:436:24: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:436:50: E202: whitespace before ')'
ERROR: plugins/module_utils/job.py:449:23: E275: missing whitespace after keyword
ERROR: plugins/module_utils/job.py:449:24: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:449:43: E202: whitespace before ')'
ERROR: plugins/module_utils/job.py:457:27: E275: missing whitespace after keyword
ERROR: plugins/module_utils/job.py:457:28: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:457:63: E202: whitespace before ')'
ERROR: plugins/module_utils/job.py:462:23: E275: missing whitespace after keyword
ERROR: plugins/module_utils/job.py:462:24: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:462:45: E202: whitespace before ')'
ERROR: plugins/module_utils/job.py:472:15: E275: missing whitespace after keyword
ERROR: plugins/module_utils/job.py:472:16: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:472:41: E202: whitespace before ')'
ERROR: plugins/module_utils/job.py:476:15: E275: missing whitespace after keyword
ERROR: plugins/module_utils/job.py:476:16: E201: whitespace after '('
ERROR: plugins/module_utils/job.py:476:40: E202: whitespace before ')'
ERROR: plugins/modules/zos_data_set.py:980:17: E222: multiple spaces after operator
ERROR: plugins/modules/zos_job_submit.py:702:21: E222: multiple spaces after operator
ERROR: plugins/modules/zos_job_submit.py:710:76: E262: inline comment should start with '# '
See documentation for help: https://docs.ansible.com/ansible-core/2.15/dev_guide/testing/sanity/pep8.html
Running sanity test "pslint"
Running sanity test "pylint"
ERROR: Found 1 pylint issue(s) which need to be resolved:
ERROR: plugins/module_utils/job.py:426:18: anomalous-backslash-in-string: Anomalous backslash in string: '\$'. String constant might be missing an r prefix.
See documentation for help: https://docs.ansible.com/ansible-core/2.15/dev_guide/testing/sanity/pylint.html
Running sanity test "replace-urlopen"
Running sanity test "runtime-metadata"
Running sanity test "shebang"
Running sanity test "shellcheck"
Running sanity test "symlinks"
Running sanity test "use-argspec-type-path"
Running sanity test "use-compat-six"
Running sanity test "validate-modules"
Running sanity test "yamllint"
FATAL: The 3 sanity test(s) listed below (out of 32) failed. See error output above for details.
import --python 3.10
pep8
pylint

```

File: changelogs/fragments/642-escape-dsnames.yml
Comment (fernandofloresg) [inline changelogs/fragments/642-escape-dsnames.yml:1]: Got a parsing error of this yaml 

```

(AC2.14PY3.9) fernandoflores@Fernandos-MacBook-Pro ibm_zos_core % antsibull-changelog lint
/Users/fernandoflores/Documents/IBM/venv/AC2.14PY3.9/lib/python3.9/site-packages/urllib3/__init__.py:34: NotOpenSSLWarning: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(
changelogs/fragments/642-escape-dsnames.yml:0:0: yaml parsing error


```

I think the correct format should be

```

minor_changes:
- module_utils/job.py - new tool available called _dsname_escape.
    Routine called _dsname_escape will ensure a dataset is escaped for cmd line compatibility.
    It will also raise an error if any data set name rules are violated.
    (https://github.com/ansible-collections/ibm_zos_core/pull/1020).


bugfixes:
- zos_data_set - Uses new _dsname_escape to make sure data set names are usable.
  (https://github.com/ansible-collections/ibm_zos_core/pull/1020).

```

Also consider tweaking the phrasing, consider this will be read by our users as part of the release notes, this enhancement you just did is a great addition, as I understand this enables a lot of new characters for data set names in our collection that are accepted by z/OS. I use this as a guide : https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment

For bugfixes (was wrongly tagged as `bugfix`): Use past tense to describe the problem and present tense to describe the fix.

I don't think we are obligated to follow that but I really like that style.

File: plugins/module_utils/job.py [hunk 1]
```diff
@@ -393,3 +393,109 @@ def _ddname_pattern(contents, resolve_dependencies):
             )
         )
     return str(contents)
+
+
+def FixDSName( name_in ):
```
Comment (fernandofloresg) [inline plugins/module_utils/job.py:398]: I see the C influence in this fragment of code, just change the function name to adhere to this PEP "Function names should be lowercase, with words separated by underscores as necessary to improve readability." https://peps.python.org/pep-0008/#function-and-variable-names

File: plugins/module_utils/job.py [hunk 2]
```diff
@@ -393,3 +393,109 @@ def _ddname_pattern(contents, resolve_dependencies):
...
+        return( None, "Too few segments" )
+
+    # DS Name may not end with a period
+    if c == '.':
+        return( None, "Ends with a dot" )
+
+    return (ret_str.join(result), None)
+
+
+def _dsname_escape(dsname_in):
```
Comment (fernandofloresg) [inline plugins/module_utils/job.py:481]: Also from the python style guide:  "_single_leading_underscore: weak “internal use” indicator. E.g. from M import * does not import objects whose names start with an underscore."

So it think FixDSName should have the single leading underscore and dsname_escape should not, since you import dsname_escape in zos_data_set module.

File: plugins/module_utils/job.py [hunk 3]
```diff
@@ -393,3 +393,109 @@ def _ddname_pattern(contents, resolve_dependencies):
...
+    """Takes in a string for dataset name type arguments
+
```
Comment (fernandofloresg) [inline plugins/module_utils/job.py:483]: Thank you for adding this docstring 👍

File: plugins/module_utils/job.py
Comment (fernandofloresg) [inline plugins/module_utils/job.py:1]: I think I recall that we talked about how to handle the data set name cleanup in terms of the workflow. Cleaning the name from the start of the module operation would be troublesome right? since that could intervene with other validations and when returning the error message to the user they may see the wrong name and will be confused.

But do you think this function would make more sense in the BetterArgParser or data_set.py module utils ? I don't think it conceptually belongs to job.py

File: plugins/modules/zos_data_set.py [hunk 4]
```diff
@@ -971,18 +973,26 @@ def perform_data_set_operations(name, state, **extra_args):
...
-        changed = DataSet.ensure_absent(name, extra_args.get("volumes"))
-    elif state == "absent" and extra_args.get("type") == "MEMBER":
-        changed = DataSet.ensure_member_absent(name, extra_args.get("force"))
-    elif state == "cataloged":
-        changed = DataSet.ensure_cataloged(name, extra_args.get("volumes"))
-    elif state == "uncataloged":
-        changed = DataSet.ensure_uncataloged(name)
+
+    # _dsname_escape will make sure original_name is properly escaped and follows dsname rules
+    # if original_name violates rules, name will be None
```
Comment (fernandofloresg) [inline plugins/modules/zos_data_set.py:978]: Is this the case? Shouldn't a wrong name raise a ValueError ?

File: plugins/module_utils/job.py [hunk 5]
```diff
@@ -393,3 +393,109 @@ def _ddname_pattern(contents, resolve_dependencies):
...
+        ValueError: When contents is invalid argument type
+
+    Returns:
+        str -- escaped string of dsname_in
+    """
+    result, errmsg = FixDSName(dsname_in)
+
+    if errmsg:
+        raise ValueError(
+            'Error processing "{0}". as data set name: "{1}"'.format(
```
Comment (fernandofloresg) [inline plugins/module_utils/job.py:497]: We don't need the point after "{0}"

File: plugins/modules/zos_job_submit.py [hunk 6]
```diff
@@ -698,13 +699,15 @@ def submit_src_jcl(module, src, src_name=None, timeout=0, hfs=True, volume=None,
...
+        job_submitted = jobs.submit(clean_name, wait, None, **kwargs)
 
-        # Introducing a sleep to ensure we have the result of job sumbit carrying the job id
+        # Introducing a sleep to ensure we have the result of job submit carrying the job id
         while (job_submitted is None and duration <= timeout):
+            sleep(0.5)
             current_time = timer()
             duration = round(current_time - start_time)
-            sleep(0.5)
+            job_submitted = jobs.submit(clean_name, wait, None, **kwargs)  ### Not sure, but think this is needed
```
Comment (fernandofloresg) [inline plugins/modules/zos_job_submit.py:710]: So apparently this while is just waiting to fail if job_submitted is None. Is it correct that we try to submit a new job?  I don't think so, I would leave as it is since we know big changes are coming to ZOAU 1.3 in terms of job submit and job fetch

**Review by fernandofloresg — CHANGES_REQUESTED** [review]

File: changelogs/fragments/642-escape-dsnames.yml
Comment (richp405) [reply changelogs/fragments/642-escape-dsnames.yml:1]: Fernando, I agree and will touch this up

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:1]: I will re-run... not sure how this got through the pipeline with whitespace errors.

File: plugins/module_utils/job.py (hunk 1 above)
Comment (richp405) [reply plugins/module_utils/job.py]: just made that change

File: plugins/module_utils/job.py (hunk 2 above)
Comment (richp405) [reply plugins/module_utils/job.py]: yep... changed to _fix_ds_name, and dsname_escape...

File: plugins/module_utils/job.py
Comment (richp405) [reply plugins/module_utils/job.py:1]: I see them running locally... I've addressed them... need to re-run pipeline and see if it comes back clean

**Comment by ddimatos** [comment]
@richp405 - i started to review this but decided to put that on hold till I have a better understanding and comments below might help that. Albeit, I am making assumptions below, but discussion will help me. 

1) Does this implementation bypass the better argparser; or let me rephrase it differently, the method `_fix_ds_name` is isolated to usage for `module_utils` `dataset.py`, `zos_data_set` and `zos_job_submit`, while I see escaping a problem for the entire collection, why is is isolated to these code paths? I see escaping an issue for any module taking a data set name as input, eg zos_blockinfile, zos_lineinfile, zos_fetch, etc 

2) In my mind, I would have leveraged the better arg parser since it is the one consistent tool we have that validates all inputs used by nearly all modules, why haven't we written a better_arg_parser routine or simply enhanced the the validation function [_data_set_type](https://github.com/ansible-collections/ibm_zos_core/blob/87f1523c8fbb3ba707ce2aca71a0cd0b5093986d/plugins/module_utils/better_arg_parser.py#L316C9-L316C23)? 

3) I only did a quick prototype to disable the better arg parser to allow the special chars to flow through in my comment [here](https://github.com/ansible-collections/ibm_zos_core/issues/724#issuecomment-1496295660) , **in my mind I was under the impression what we should be doing is when we see one of the special chars, we escape it.** 

For example, a user can submit this in data set name, src=ANSIBLE.MEMBER.TEST(SUB$CL)' and what we will do is convert that with our better argparser with an updated name to be used internally in the code that would be this `ANSIBLE.MEMBER.TEST(SUB\$CL)` , notice the `\$` being added because that is what ZOAU expects. What I am seeing in this code are things like this in the PR which implies to me are going to be allowing users to escape data set names when they provide us the source, eg `src=ANSIBLE.MEMBER.TEST(SUB\\$CL)' <--- I think this should **not** be the approach because permitted alphabetic characters are digits (0-9), a hyphen (-), #  or the special characters $, #, @. A user passing a an escape as part of the src data set name would invalidate that and then we are making it more confusing for users, its better just to accept the special chars and if they need to be escaped, we escape them instead of putting the onus on users. 
```
        # discard incoming escapes
        if c == "\\":
```

4) I see inconsistency in Data set name length, while yes 44 is permitted but our better arg parser seems to have allowed 3 Qualifiers each being a max of 8 chars (3 * 8 = 24) + what ever HLQ the system may add, usually a USERNAME, yet you are allowing 44. While [I don't think 44 is incorrect](https://www.ibm.com/docs/en/zos/2.4.0?topic=conventions-data-set-naming-rules) but we have a [discrepancy](https://github.com/ansible-collections/ibm_zos_core/blob/87f1523c8fbb3ba707ce2aca71a0cd0b5093986d/plugins/module_utils/better_arg_parser.py#L330) that needs to be understood.

**Comment by richp405** [comment]
addressed in prior ticket.
